Refactor + select support

BitPhoenix opened the pull request:
Hi Omar I'm BitPhoenix, it's a pleasure to meet you. I wanted to start off by thanking you for creating such an awesome package! It's very useful and has helped me a lot. I really appreciate the hard work you continue to put into it.

My apologies that the refactor and fix are in one pull review, and that the commits are larger than normal. I'll make sure to separate them in the future. 

Since this is quite a big pull request, please let me know if you'd like to hop on a voice chat to go through the changes together, in case that would help make things easier. 

### References
- Support select: https://github.com/omar-dulaimi/prisma-trpc-generator/issues/13
- Support [select relation fields](https://www.prisma.io/docs/concepts/components/prisma-client/select-fields#include-relations-and-select-relation-fields)
- Support [select _count](https://www.prisma.io/docs/concepts/components/prisma-client/select-fields#relation-count)

### Description
I started out by trying to fix the missing select issue referenced here: https://github.com/omar-dulaimi/prisma-trpc-generator/issues/13

To help me better understand the codebase, I started refactoring the `prisma-generator.ts` and `transformer.ts` files as I read through the code to get hands on and aid in understanding. Pretty small things like:
- Extracting grouped logic into functions in `prisma-generator.ts`
- Updating some variable / function names for clarity / consistency
- Reorganizing function order in `transformer.ts` to decrease [vertical distance ](https://mosquitolwz.medium.com/clean-code-quotes-5-formatting-b7bdd4a828d6#:~:text=appear%20vertically%20dense.-,Vertical%20Distance,-Concepts%20that%20are)
- Extracting missing mongo type generation into a separate file + functions

After I finished performing the above changes and understood the codebase, I added functions to `src/helpers/helpers.ts` responsible for generating the types needed to support `select`. I plan on working on `include` next and will likely update the file / folder structure in a future pull request to represent that this group of functions is for adding the select related types (similar to how we now have a file dedicated to mongo missing type generation logic). 

The screenshots below show the generated zod schema which will help a lot to explain what the end goal was of the code that was added to `src/helpers/helpers.ts`. The short version is:
- support select
- support [select relation fields](https://www.prisma.io/docs/concepts/components/prisma-client/select-fields#include-relations-and-select-relation-fields)
- support [select _count](https://www.prisma.io/docs/concepts/components/prisma-client/select-fields#relation-count)

Appropriate queries have a `select` property (Unique example is pictured)
![Screen Shot 2022-10-28 at 3 32 10 PM](https://user-images.githubusercontent.com/46142503/198717786-d871a022-525e-4b0e-8760-624ccf3aab2e.png)

ModelSelect schema is generated. If the model has a **many** relation with another model, we include a union with the find many schema for that model. We also include the `_count` property
![image](https://user-images.githubusercontent.com/46142503/198718415-7eb389ec-ea18-46f1-983a-ff4c28eb6c7d.png)

If the model has a **single** relation with another model, we union with args object schema for that model. If the model does not have a many relation with another model, _count will not be present.
![Screen Shot 2022-10-28 at 3 37 48 PM](https://user-images.githubusercontent.com/46142503/198718687-eab966b3-6683-4ec3-b9c1-4ac5694d8b71.png)

omar-dulaimi commented on src/prisma-generator.ts:
```diff
@@ -1,123 +1,114 @@
 
-  const inputObjectTypes = prismaClientDmmf.schema.inputObjectTypes.prisma;
-  const rawOpsMap: { [name: string]: string } = {};
+async function handleGeneratorOutputValue(generatorOuputValue: EnvValue) {
```
typo in `generatorOuputValue`
should be `generatorOutputValue`

missing `t`

BitPhoenix replied:
Great catch! Fixed ✅

omar-dulaimi commented on src/transformer.ts:
```diff
@@ -291,53 +284,131 @@ export default class Transformer {
 
-        return value.trim();
-      });
+  checkIsModelQueryType(type: string) {
```
This function is used in `generateSchemaImports`, and the cases there only `findMany`, `enum` and `object schemas`. My question is how does it handle others like findUnique?

I might be missing something about `checkIsModelQueryType`.

BitPhoenix replied:
Great question! First I will try to write out some background info for useful context:

When generating `{Model}Select` in `generateModelSelectInputObjectTypes`, we have the following line that is inspired by the generated Prisma types for select:

`type: isList ? `${type}FindManyArgs` : `${type}Args`,`

Examples:
On UserSelect: `posts: boolean | PostFindManyArgs`
On PostSelect: `author: boolean | UserArgs`

For`{type}Args`, this is a new zod schema we will have to generate, which we account for in `generateModelArgsInputObjectTypes`

However for `{type}FindManyArgs`, our `{Model}FindManySchema` at `findMany{Model}.ts` is the perfect fit, and we do not have to generate another file.

The tricky part is that up until now `generateSchemaImports` did not need to support importing our top level zod query schemas, which is why `checkIsModelQueryType` and `resolveModelQuerySchemaName` came into existence, and why we `import { ${name}Schema } from '../enums/${name}.schema'` 

Currently we only have the `FindManyArgs -> 'findMany'` mapping in `checkIsModelQueryType`, as this is the first time that: one of the fields in an input type object we've pushed is a top level query schema (findMany{Model}). 

And lastly now that we've generated the `{Model}Select` schemas, we can add `select` to the top level query zod schemas that support select in `generateModelSchemas`. `z.object({ select: ${modelName}SelectObjectSchema.optional(),` 🙂

omar-dulaimi replied:
@BitPhoenix Thank you for the detailed explanation, really drove the point home.

omar-dulaimi reviewed (changes requested):
Hey @BitPhoenix 
I'm glad you like this project and find it useful.

Let me start by saying that I don't know what is better, the description of the PR, the code refactor or the feature implementation!

Not only you did all that, but you also taught me a few things in the process; like vertical distance, the importance of high cohesion and separation of concerns at a practical level.

Thank you for that.

I left you just one small note(a typo) and a question in the review.

Also, if you could, maybe add a flag that defaults to false, that controls outputting `select` files. This way it would maintain the current behavior and the current amount of files people expect.

Other than that, I think you did great. I appreciate the time and effort you put into this.

Shahidul1004 commented:
Hi @BitPhoenix,
I am getting error for circular relations in `select`.
I integrated this PR to my tester app. I am getting full type support in vscode but APIs are not working. See the schemas. If I break the circular relation, then the APIs will work.
![Screenshot from 2022-11-02 00-38-33](https://user-images.githubusercontent.com/62707100/199326470-e3f454b6-dd60-4c66-97d8-d93b24b0c9b0.png)
![Screenshot from 2022-11-02 00-44-25](https://user-images.githubusercontent.com/62707100/199326475-15fc7000-8826-4df8-9008-4b06cc5d80f7.png)

BitPhoenix commented:
> Hi @BitPhoenix, I am getting error for circular relations in `select`. I integrated this PR to my tester app. I am getting full type support in vscode but APIs are not working. See the schemas. If I break the circular relation, then the APIs will work. ![Screenshot from 2022-11-02 00-38-33](https://user-images.githubusercontent.com/62707100/199326470-e3f454b6-dd60-4c66-97d8-d93b24b0c9b0.png) ![Screenshot from 2022-11-02 00-44-25](https://user-images.githubusercontent.com/62707100/199326475-15fc7000-8826-4df8-9008-4b06cc5d80f7.png)

Hey Shahidul1004 nice to meet ya! Thank you for testing this out and submitting the very helpful information. 

Interestingly when I try it out on my end I don't have any issues, I suspect it's working on my end is a side effect of node's behavior for supporting circular imports:
- [Node supports circular imports](https://spin.atomicobject.com/2018/06/25/circular-dependencies-javascript/#:~:text=Node.js%20does%20support%20circular%20require/import%20statements%20between%20modules)
- [Node.js documentation](https://nodejs.org/api/modules.html#modules_cycles)

Will attach screenshots of it working on my end below as a sanity check 

![image](https://user-images.githubusercontent.com/46142503/199383706-93bf45ad-256f-4d4e-a13e-298bbe3d1333.png)

<img width="553" alt="image" src="https://user-images.githubusercontent.com/46142503/199383782-d196d0d4-78de-4038-abe0-9603824bf5e8.png">

Do you happen to have your code pushed to a branch somewhere where the three of us can test it out / reproduce the error to make it easier to work on a fix? I don't get the error in my current environment but can definitely see this being an issue in other ones. So just want to make sure I can be in one where the error occurs for fixing / testing any fixes 🙂 

I'm curious if there's any difference if you manually update the generated schema to this:

```
export const PostFindManySchema = z.object({
  where: PostWhereInputObjectSchema.optional(),
  orderBy: PostOrderByWithRelationInputObjectSchema.optional(),
  cursor: PostWhereUniqueInputObjectSchema.optional(),
  take: z.number().optional(),
  skip: z.number().optional(),
  distinct: z.array(PostScalarFieldEnumSchema).optional(),
})
.strict();
Object.assign(PostFindManySchema,  PostFindManySchema.extend({ select: PostSelectObjectSchema.optional() }));
```
Curious if you'll get the same error with this snippet. If no error this might be a potential fix, the only downside being that the select is not available in the type if you hover over the imported PostFindManySchema. Note that select is actually present and working on the exported schema - can be verified by parsing an object containing select first, then commenting out Object.assign and testing again. It's just not shown on the type when hovering with mouse, but is actually part of the exported / imported zod schema.

If this doesn't make a difference and we get the same error, we'll have to brainstorm a bit more on how to resolve this.

BitPhoenix commented:
@omar-dulaimi Update on most recents commits:
- Merged in @Revitate aggregate support
- Extracted aggregate support to helpers/aggregate-helpers.ts
- Renamed helpers.ts to select-helpers.ts
- Now calling mongodb input object type generation / pushing from `addMissingInputObjectTypes`
- Fixed function parameter typo

Before merging: note that discussion / investigation with @Shahidul1004 is pending to resolve the issue he is experiencing

BitPhoenix commented:
> Hey @BitPhoenix I'm glad you like this project and find it useful.
> 
> Let me start by saying that I don't know what is better, the description of the PR, the code refactor or the feature implementation!
> 
> Not only you did all that, but you also taught me a few things in the process; like vertical distance, the importance of high cohesion and separation of concerns at a practical level.
> 
> Thank you for that.
> 
> I left you just one small note(a typo) and a question in the review.
> 
> Also, if you could, maybe add a flag that defaults to false, that controls outputting `select` files. This way it would maintain the current behavior and the current amount of files people expect.
> 
> Other than that, I think you did great. I appreciate the time and effort you put into this.

Thank you so much for the kind words 🙏  And for the quick review!

I'm happy to hear that I was able to share a few things, I learned a lot from you about how prisma generators work! Very grateful that I got to give back in some way.

"Also, if you could, maybe add a flag that defaults to false, that controls outputting select files. This way it would maintain the current behavior and the current amount of files people expect."

Nice idea - I'll work on this one next!

Shahidul1004 commented:
Hi @BitPhoenix 
Hope you are doing great.
I got a solution. We need to call `PostSelectObjectSchema` as a lazy function inside of `PostFindManySchema`. Because we are calling `PostSelectObjectSchema` inside `PostFindManySchema`, but `PostSelectObjectSchema` is defined later.

This circular relation exists in the `Post model` but not in the `User model`. So we can call `{Model}SelectObjectSchema` as a lazy function inside of `{Model}FindManySchema` if the `{Model}` has `many-to-one` relation with other models. But it would be fine, if we always lazily call `{Model}SelectObjectSchema` inside `{Model}FindManySchema`.
![Screenshot from 2022-11-02 16-49-39](https://user-images.githubusercontent.com/62707100/199471123-1810c29d-4194-4b88-9290-e5d2645f9876.png)

Shahidul1004 commented:
Hey @BitPhoenix 
I was trying to add `include` support for this repo and, it seems like I have succeeded. I was wondering if you are also working on the same thing. If that's not the case then I might create a PR for `include` support.

BitPhoenix commented:
> Hi @BitPhoenix Hope you are doing great. I got a solution. We need to call `PostSelectObjectSchema` as a lazy function inside of `PostFindManySchema`. Because we are calling `PostSelectObjectSchema` inside `PostFindManySchema`, but `PostSelectObjectSchema` is defined later.
> 
> This circular relation exists in the `Post model` but not in the `User model`. So we can call `{Model}SelectObjectSchema` as a lazy function inside of `{Model}FindManySchema` if the `{Model}` has `many-to-one` relation with other models. But it would be fine, if we always lazily call `{Model}SelectObjectSchema` inside `{Model}FindManySchema`. ![Screenshot from 2022-11-02 16-49-39](https://user-images.githubusercontent.com/62707100/199471123-1810c29d-4194-4b88-9290-e5d2645f9876.png)

Fantastic solution and explanation! Added a commit with your solution: https://github.com/omar-dulaimi/prisma-zod-generator/pull/28/commits/12b4b09f6e29e00e9d69284f89f9fbebd3b8ad3c

Added co-authored by line for you in the commit message but not sure if it worked properly

BitPhoenix commented:
> Hey @BitPhoenix I was trying to add `include` support for this repo and, it seems like I have succeeded. I was wondering if you are also working on the same thing. If that's not the case then I might create a PR for `include` support.

That's fantastic! I was going to work on include later this week so nope all good! I'll add the feature flag request today that omar requested for select generation and hopefully we can get PR one merged in and then we can look over your include PR next 🎊 

You probably already did this but just in case, one suggestion I have is to put include function code under `helpers/include-helpers.ts` and then call the function from `helpers.ts addMissingInputObjectTypes`, basically just keeping the same convention this pr introduces for organized missing input object type generation logic: so far we have aggregate, select, mongodb and with yours we'd also have a dedicated file for include 🙂

BitPhoenix commented:
@omar-dulaimi Small update: added a commit with @Shahidul1004's fix for the circular relation. Last part is to add the flag support for select generation which I will add today so that we can tackle @Shahidul1004's include functionality PR next

BitPhoenix commented:
@omar-dulaimi Most recent commit contains isGenerateSelect generator config option that controls whether Select schemas are generated and whether the select property is available on query schemas (defaults to false)

@Shahidul1004 Might be worth reviewing that commit as I imagine Omar might potentially want include to be flag controlled as well

BitPhoenix commented:
Ready for re-review ✅

omar-dulaimi reviewed (changes requested):
Thank you guys for your hard work on this PR so far. I didn't find anything major to comment on in the review, since everything was taken care of. Great catch though from @Shahidul1004 for the cyclical ref issue, it's kind of an annoying issue I faced in another generator. Glad that you guys managed to solve it with Zod.

BitPhoenix commented on README.md:
```diff
@@ -108,15 +108,17 @@ app.post('/blog', async (req, res, next) => {
 generator zod {
   provider   = "prisma-zod-generator"
   output     = "./generated-zod-schemas"
+  isGenerate = true
```
Good catch! Fixed ✅

BitPhoenix commented on src/transformer.ts:
```diff
@@ -535,26 +643,37 @@ export default class Transformer {
+  resolveSelectImportAndZodSchemaLine(modelName: string) {
+    const selectImport = Transformer.isGenerateSelect
+      ? `import { ${modelName}SelectObjectSchema } from './objects/${modelName}Select.schema'`
+      : undefined;
```
Thank you 🙏 Great idea on keeping all 3 defaults consistent, adjusted to empty string ✅

BitPhoenix commented:
Ready for re-review: ✅ 
- Updated README isGenerate -> isGenerateSelect
- Updated default value to be `''` instead of `undefined`

omar-dulaimi commented:
Released in https://github.com/omar-dulaimi/prisma-zod-generator/releases/tag/0.6.0

Thanks for your efforts

BitPhoenix commented:
@omar-dulaimi Thank you for all of your time and insightful comments on this pull request! 🙏
